### PR TITLE
fix: Fix quick cast not working with crafted weapons [skip ci]

### PR DESCRIPTION
**Pull request by ShadowCat117**
https://github.com/user-attachments/assets/6a18e0cb-a33b-47bc-82f8-b48a1ff3ec9e

### CODE REVIEW

**Review by kristofbolyai — APPROVED** [review]

File: common/src/main/java/com/wynntils/features/combat/QuickCastFeature.java [hunk 1]
```diff
@@ -124,17 +126,30 @@ private void tryCastSpell(SpellUnit a, SpellUnit b, SpellUnit c) {
...
-                return;
-            } else if (!gearItemOpt.get().meetsActualRequirements()) {
+                Optional<CraftedGearItem> craftedGearItemOpt = Models.Item.asWynnItem(heldItem, CraftedGearItem.class);
+
+                if (craftedGearItemOpt.isEmpty()) {
+                    sendCancelReason(Component.translatable("feature.wynntils.quickCast.notAWeapon"));
+                    return;
+                }
+
+                isArcher = craftedGearItemOpt.get().getRequiredClass() == ClassType.ARCHER;
```
Comment (magicus) [inline common/src/main/java/com/wynntils/features/combat/QuickCastFeature.java]: Don't we have a common interface for gear item and crafted gear items that can supply getRequiredClass()? If not, maybe we should create one? "ClassableItemProperty" or whatever...

Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/features/combat/QuickCastFeature.java]: Not currently but I can do it here if that would be better

**Comment by Twisted-Code** [comment]
I see you noticed this issue already, so I won't bother opening an issue for it. I will say, however, that I'm a little disappointed you didn't catch this one before it was released. Had to switch weapons temporarily on my assassin, although I suppose I could've also downgraded off the bleeding edge.
Could you please consider adding same unit testing to prevent stuff like this from slipping through in the future? I also think you are onto something with making an interface.

**Comment by ShadowCat117** [comment]
> I see you noticed this issue already, so I won't bother opening an issue for it. I will say, however, that I'm a little disappointed you didn't catch this one before it was released. Had to switch weapons temporarily on my assassin, although I suppose I could've also downgraded off the bleeding edge. Could you please consider adding same unit testing to prevent stuff like this from slipping through in the future? I also think you are onto something with making an interface.

A config option to ignore checking for valid weapons was added exactly for this reason, so that if our gear checking breaks again (like it did with 2.1) then that can be toggled temporarily so that it will still work. As for not catching it, that's my fault for forgetting we handle regular and crafted gear separately.

And for the unit testing, due to the nature of the mod there isn't really anything we can do to setup automatic testing, we have to do it all manually. As it requires you to log onto Wynn and perform actions

**Comment by Twisted-Code** [comment]
Oh, thanks for the information about the config option. Didn't know about that.
You wouldn't be able to extract the validity checking to a separate method and then feed it a few example itemstacks? Even if you can't make it fully automatic with CI actions, would be nice if, while manually testing, some code in the background warns you if something doesn't seem to be behaving correctly. Obviously if CI could handle it that would be the ideal way to test but I know TDD is difficult for games, game mods, etc.; a lot of moving parts.

**Comment by magicus** [comment]
@Twisted-Code We would love to have more automated testing, and this has definitely been considered. But it would require a ton of work to setup a "mock" Wynntils server to test against, and nobody have been inclined to put in that amount of work. Feel free to do it yourself, if you want to.

File: common/src/main/java/com/wynntils/features/combat/QuickCastFeature.java (hunk 1 above)
Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/features/combat/QuickCastFeature.java]: Not added it to UnknownGearItem as we don't have any class info there

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/combat/QuickCastFeature.java]: UnknownGearItem does have `requirements`, containing class info.

Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/features/combat/QuickCastFeature.java]: So it does, I missed that 😅

**Comment by Twisted-Code** [comment]
Yeah that's fair, I was made aware in a previous issue that this project is a bit short on contributor time. I'd like to find some ways to help with that which don't require getting too far outside my comfort zone, but preferably more than just posting issues like I have been.
Back to test automation, what exactly has been discussed so far? I make no promises, at all, but I might™ look into it if I can get an adequate idea of the requirements to be met.

**Comment by magicus** [comment]
Nothing has been really discussed. I guess what we really would need is a scriptable minecraft server that implement all extended functionality that wynncraft implements. But that is like just about the same amount of work as the entire wynntils project... Short of that, I don't really know what would be helpful. Not much, unfortunately.

**Pull request merged**